File: C++/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        vector<int> digits;
        int val = 0;
        int carry = 0;
        
        while (l1 || l2) {
            val = carry + (l1 ? l1->val : 0) + (l2 ? l2->val : 0);
            carry = val / 10;
            val = val % 10;
            
            digits.push_back(val);
            
            l1 = l1 ? l1->next : l1;
            l2 = l2 ? l2->next : l2;
        }
        
        if (carry) {
            digits.push_back(carry);
        }
        
        ListNode *result = nullptr;
        
        while (digits.size() > 0) {
            result = new ListNode(digits.back(), result);
            digits.pop_back();
        }
        
        return result;
    }
};

File: Kotlin/add-two-numbers.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {

        // add digits in reverse order
        val s = ArrayList<Int>()
        var n1 = l1
        var n2 = l2
        var carry = 0
        while (n1 != null || n2 != null ) {
            var x = carry
            if (n1 != null) {
                x += n1.`val`
                n1 = n1.next
            }
            if (n2 != null) {
                x += n2.`val`
                n2 = n2.next
            }
            carry = x / 10
            s.add(x % 10)
        }
        if (carry > 0) {
            s.add(carry)
        }

        // build result by reversing s
        var result: ListNode? = null
        for (x in s.reversed()) {
            val n = ListNode(x)
            n.next = result
            result = n
        }

        return result
    }
}